docs/source/general/data.rst

```diff
@@ -2,11 +2,371 @@
+        ):
+
+            preprocess = preprocess or cls.preprocess_cls()
+
+            # {stage}_load_data_input will be given to your
+            # ``Preprocess`` ``{stage}_load_data`` function.
+            return cls.from_load_data_inputs(
+                    train_load_data_input=train_folder,
+                    val_load_data_input=val_folder,
+                    test_load_data_input=test_folder,
+                    predict_load_data_input=predict_folder,
+                    preprocess=preprocess,  # DON'T FORGET TO PASS THE CREATED PREPROCESS
+                    **kwargs,
+                )
+
+    # When structuring your code, it is possible to pass the ``preprocess`` directly to the DataModule ``classmethod``.
+    preprocess = ImageClassificationPreprocess()
+
+    dm = ImageClassificationDataModule.from_folders(
+        "./data/train",
+        "./data/val",
+        "./data/test",
+        "./data/predict",
+        preprocess=preprocess)
+
+    model = ImageClassifier(...)
+    trainer = Trainer(...)
+
+    trainer.fit(model, dm)
+
+
+
+*************
+API reference
+*************
+
+.. _preprocess:
+
+Preprocess
+__________
+
+.. autoclass:: flash.data.process.Preprocess
+    :members:
+
+
+----------
+
+.. _postprocess:
+
+Postprocess
+___________
+
+
+.. autoclass:: flash.data.process.Postprocess
+    :members:
+
+
+----------
+
 .. _datapipeline:
 
 DataPipeline
-------------
+____________
+
+.. autoclass:: flash.data.data_pipeline.DataPipeline
+    :members:
+
+----------
+
+DataModule
+__________
+
+.. autoclass:: flash.data.data_module.DataModule
+    :members:
+
+
+******************************
+How it works behind the scenes
+******************************
+
+Preprocess
+__________
+
+.. note:: The ``load_data`` and ``load_sample`` will be used to generate an AutoDataset object.
+
+Here is the ``AutoDataset`` pseudo-code.
+
+Example::
+
+    from pytorch_lightning.trainer.states import RunningStage
+
+    class AutoDataset
+        def __init__(
+            self,
+            data: Any,
+            load_data: Optional[Callable] = None,
+            load_sample: Optional[Callable] = None,
+            data_pipeline: Optional['DataPipeline'] = None,
+            running_stage: Optional[RunningStage] = None
+        ) -> None:
+
+            self.preprocess = data_pipeline._preprocess_pipeline
+            self.preprocessed_data: Iterable = self.preprocess.load_data(data)
+
+        def __getitem__(self, index):
+            return self.preprocess.load_sample(self.preprocessed_data[index])
+
+        def __len__(self):
+            return len(self.preprocessed_data)
+
+.. note::
+
+    The ``pre_tensor_transform``, ``to_tensor_transform``, ``post_tensor_transform``, ``collate``,
+    ``per_batch_transform`` are injected as the ``collate_fn`` function of the DataLoader.
+
+Here is the pseudo code using the preprocess hooks name.
+Surely, Flash will take care of calling the right hooks for each stage.
```

Comment:
Flash takes care of calling the right hooks for each stage.

Reply:
Done !